esp/bindings/http/platform/httpservice.cpp

```diff
@@ -974,3 +875,552 @@ int CEspHttpServer::onGet()
+    {
+        ESPLOG(LogMin, "ESP password expired for %s", authReq.ctx->queryUserId());
+        handlePasswordExpired(sessionAuth);
+        return;
+    }
+
+    if (!sessionAuth)
+    {
+        ESPLOG(LogMin, "Authentication failed: send BasicAuthentication.");
+        m_response->sendBasicChallenge(authReq.authBinding->getChallengeRealm(), true);
+        return;
+    }
+
+    ESPLOG(LogMin, "Authentication failed: call askUserLogin.");
+    if (sessionID)
+        askUserLogin(authReq.authBinding, *sessionID);
+    else
+        askUserLogin(authReq.authBinding, createHTTPSession(authReq, nullptr, nullptr));
+}
+
+void CEspHttpServer::askUserLogin(EspHttpBinding* authBinding, unsigned sessionID)
+{
+    VStringBuffer sessionIDStr("%u", sessionID);
+    addCookie(authBinding->querySessionIDCookieName(), sessionIDStr.str(), authBinding->getSessionTimeoutSeconds());
+    addCookie(SESSION_AUTH_COOKIE, "1", 0); //time out when browser is closed
+    m_response->redirect(*m_request, authBinding->queryLoginURL());
+}
+
+void CEspHttpServer::send200OK()
+{
+    m_response->setContentType("text/html; charset=UTF-8");
+    m_response->setStatus(HTTP_STATUS_OK);
+    m_response->send();
+}
+
+unsigned CEspHttpServer::createHTTPSession(EspAuthRequest& authReq, IPropertyTree* domainSessions, const char* _loginURL)
+{
+    CDateTime now;
+    now.setNow();
+    time_t createTime = now.getSimple();
+
+    StringBuffer peer;
+    VStringBuffer idStr("%s_%ld", m_request->getPeer(peer).str(), createTime);
+    unsigned sessionID = hashc((unsigned char *)idStr.str(), idStr.length(), 0);
+    ESPLOG(LogMax, "New sessionID <%d> at <%ld> in createHTTPSession()", sessionID, createTime);
+
+    Owned<IPropertyTree> ptree = createPTree();
+    ptree->addProp(PropSessionNetworkAddress, peer.str());
+    ptree->addPropInt64(PropSessionID, sessionID);
+    ptree->addPropInt(PropSessionState, HTTPSS_new);
+    ptree->setPropInt64(PropSessionCreateTime, createTime);
+    ptree->setPropInt64(PropSessionLastAccessed, createTime);
+
+    if (!isEmptyString(_loginURL))
+        ptree->addProp(PropSessionLoginURL, _loginURL);
+    else
+    {
+        StringBuffer loginURL = authReq.httpPath;
+        if (authReq.requestParams && authReq.requestParams->hasProp("__querystring"))
+            loginURL.append("?").append(authReq.requestParams->queryProp("__querystring"));
+        if (!loginURL.isEmpty() && streq(loginURL.str(), "/WsSMC/"))
+            ptree->addProp(PropSessionLoginURL, "/");
+        else
+            ptree->addProp(PropSessionLoginURL, loginURL.str());
+    }
+
+    if (domainSessions)
+    {
+        domainSessions->addPropTree(PathSessionSession, LINK(ptree));
+        return sessionID;
+    }
+
+    Owned<IRemoteConnection> conn = querySDSConnection(authReq.authBinding->queryDomainSessionSDSPath(), RTM_LOCK_WRITE, SESSION_SDS_LOCK_TIMEOUT);
```

Comment:
It would be a bit cleaner if this method was always given the domainSessions IPT
And a separate method was added when a connection was needed, e.g.:
```
unsigned CEspHttpServer::createHTTPSession(EspAuthRequest& authReq, const char* loginURL)
{
    Owned<IRemoteConnection> conn = querySDSConnection(authReq.authBinding->queryDomainSessionSDSPath(), RTM_LOCK_WRITE, SESSION_SDS_LOCK_TIMEOUT);
    IPropertyTree *newDomainSessions = conn->queryRoot()->addPropTree(PathSessionSession);
    return createHTTPSession(authReq, newDomainSessions, loginURL);
}
```